crud/common/sharding/storage_metadata_cache.lua

```diff
@@ -0,0 +1,89 @@
+local stash = require('crud.common.stash')
+local utils = require('crud.common.sharding.utils')
+
+local storage_metadata_cache = {}
+
+local FUNC = 1
+local KEY = 2
+
+local cache_data = {
+    [FUNC] = nil,
+    [KEY] = nil,
+}
+
+local ddl_space = {
+    [FUNC] = '_ddl_sharding_func',
+    [KEY] = '_ddl_sharding_key',
+}
+
+local trigger_stash = stash.get(stash.name.ddl_triggers)
+
+local function update_sharding_func_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_func_def = utils.extract_sharding_func_def(tuple)
+    cache_data[FUNC][space_name] = utils.compute_hash(sharding_func_def)
+end
+
+local function update_sharding_key_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_key_def = tuple[utils.SPACE_SHARDING_KEY_FIELDNO]
+    cache_data[KEY][space_name] = utils.compute_hash(sharding_key_def)
+end
+
+local update_hash = {
+    [FUNC] = update_sharding_func_hash,
+    [KEY] = update_sharding_key_hash,
+}
+
+local function init_cache(section)
+    cache_data[section] = {}
+
+    local space = box.space[ddl_space[section]]
+
+    local update_hash_func = update_hash[section]
+
+    -- Remove old trigger if there was some code reload.
+    -- It is possible that ddl space was dropped and created again,
+    -- so removing non-existing trigger will cause fail;
+    -- thus we use pcall.
+    pcall(space.on_replace, space, nil, trigger_stash[section])
+
+    trigger_stash[section] = space:on_replace(
+        function(_, new)
+            return update_hash_func(new)
+        end
+    )
+
+    for _, tuple in space:pairs() do
+        local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+        if cache_data[section][space_name] == nil then
+            update_hash_func(tuple)
+        end
+    end
+end
+
+local function get_sharding_hash(space_name, section)
+    if box.space[ddl_space[section]] == nil then
+        return nil
+    end
+
+    if cache_data[section] == nil then
+        init_cache(section)
+    end
```

Comment:
Here we lean on assumption that the `_ddl_sharding_*` spaces will not be re-created from scratch. It is okay (at least for now), but I would write the assumption explicitly in a comment in the code.

NB: It could be solved to some extent if we'll store space id for which the cache is created.

How nice it would be to have https://github.com/tarantool/tarantool/discussions/6800, right?

Reply:
> Here we lean on assumption that the _ddl_sharding_* spaces will not be re-created from scratch.

Yes, if one would mess up with ddl spaces directly (we already have a test doing something like this), it may break. I'll add a comment.

Reply:
Oh, there's a one safety wrapper already
![image](https://user-images.githubusercontent.com/20455996/163949520-5959b394-babc-477a-bdae-acd78a84ece9.png)

Reply:
Added a comment

Reply:
> Oh, there's a one safety wrapper already

The code on the picture will not be executed at all if we already have caches. We'll not invoke the `init_cache()` function. So I expect that the `cache_data[section] == nil` condition will be commented with assumptions under which it implements expected logic.

> Added a comment

I don't see a comment to this code block in the 'ddl: store sharding info hashes on storage' commit.

Reply:
> I don't see a comment to this code block in the 'ddl: store sharding info hashes on storage' commit.

Now I see it.